rtos/ConditionVariable.h

```diff
@@ -0,0 +1,78 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2012 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef CONDITIONVARIABLE_H
+#define CONDITIONVARIABLE_H
+
+#include <stdint.h>
+#include "cmsis_os.h"
+#include "rtos/Mutex.h"
+
+#include "platform/NonCopyable.h"
+
+namespace rtos {
+/** \addtogroup rtos */
+/** @{*/
+
+struct Waiter;
+
+/** The ConditionVariable class is used to wait for condition changes
+ * while being protected by a mutex.
+*/
+class ConditionVariable : private mbed::NonCopyable<ConditionVariable> {
+public:
+    /** Create and Initialize a ConditionVariable object */
+    ConditionVariable(Mutex &mutex);
+
+    /** Wait for a notification
+     *
+     * @param   millisec  timeout value or 0 in case of no time-out. (default: osWaitForever)
+     * @return  true if woken from notification, false if timed out.
+     *
+     * @note - The thread calling this function must be the owner of the ConditionVariable's mutex
+     */
+    bool wait(uint32_t millisec=osWaitForever);
```

Comment:
Can we have `wait_until`? It's what you're more likely to need as you're doing `wait` in a loop, so waiting for a certain time has to be done like this:

    bool wait_for_my_condition(uint32_t for_ticks)
    {
        if (my_condition) {
            return true;
        }

        uint32_t end_time = osKernelGetTickCount() + for_ticks;
        do {
            if (!cv.wait_until(end_time)) {
                 return my_condition;
            }
        } while (!my_condition);

        return true;
    }

For comparison C++11 has `wait`, `wait_until` and `wait_for`; C11 and pthreads have `wait` and `timedwait`="wait until", with no "wait for".

Reply:
Currently there aren't any C++ RTOS APIs to get the kernel tick. Because of this this couldn't be used with the C++ RTOS API alone. This would be better off as a future addition when a way to get the tick count from the C++ RTOS API is available.

Before that can be done though, the relation between OS tick and millisecond needs to get sorted out. All the existing RTOS drivers are written on the assumption that 1ms = 1 tick. If this changes in the future, waiting until an absolute time could become a lot more complex, as uint32_t ticks and uint32_t milliseconds will overflow at different points. For relative delays this is not as much of a problem since you do not need to worry about rollover. CC @pan-, @0xc0170 

I'll add the wait_for and remove the parameter from wait but wait_until will need to come in a separate PR.

Reply:
My main reservation is that I don't see how `wait_for` is really useable either without a way of getting the current time. If no C++ time API is an argument against `wait_until`, it's just as strong an argument as `wait_for`.

NSAPI already has this bug that its timed socket blocks don't actually block for the time requested, due to its pattern of repeated inner semaphore waits. But that was quite an unusual usage of semaphores.

Here, it's a requirement that users have the CV loop, and giving them only `wait_for` without the infrastructure to get their intended total wait correct seems like an invitation to "put bug here". As if `wait_for` wasn't that already - hence its total absence from pthreads and C11.

Personally I'm not a fan of RTOS APIs trying to present milliseconds - serious users will need to understand tick granularity and how to round, and specify in actual ticks, If you were to ever break the assumption 1 tick = 1 ms, then I think you'd need to have a proper tick API - you couldn't stay with just the millisecond API. (I spent a while with Seppo reworking Nanostack's timing API on that basis earlier this year - the "until" API is in native ticks only, avoiding wrapping problems).